lib/icedfrisby.js

```diff
@@ -1225,31 +1235,54 @@ class Frisby {
 
 
   _makeRequest (done) {
-    const self = this
-    let timeoutFinished = false
-    self.current.tries++
-
-    const timeoutId = setTimeout(function maxWait() {
-      timeoutFinished = true
-      if (self.current.tries < self.current.retries + 1) {
-        process.stdout.write('R')
-        self._makeRequest(done)
-      } else {
-        // In frisby it.results_ would trigger a failure for jasmine but has
-        // no effect in mocha. We need to indicate a failure for tests that
-        // reach this point.
-        const err = Error('Destination URL may be down or URL is invalid')
-        self.current.failures.push(err)
-        self._finish(done)
+    let requestAlreadyTimedOut = false
+    this.current.tries++
+
+    const maybePerformRetry = () => {
+      if (this.current.tries > this.current.retries) {
+        return false
+      }
+
+      process.stdout.write('R')
+
+      setTimeout(() => {
+        this._makeRequest(done)
+      }, this.current.retryBackoff)
+
+      return true
+    }
+
+    const handleRequestTimeout = () => {
+      requestAlreadyTimedOut = true
+
+      if (maybePerformRetry()) {
+        return
       }
-    }, self._timeout)
 
-    self.current.it(function (data) {
-      if (!timeoutFinished) {
-        clearTimeout(timeoutId)
-        self._performInspections.call(self)
-        self._invokeExpects.call(self, done)
+      let message = `Request timed out after ${this._timeout} ms`
+      if (this.current.tries > 1) {
+        message += ` (${this.current.tries} attempts)`
       }
+      this._noteExpectationFailed(Error(message))
+
+      this._finish(done)
+    }
+
+    const requestTimeoutId = setTimeout(handleRequestTimeout, this._timeout)
+
+    this.current.it(() => {
+      if (requestAlreadyTimedOut) {
+        return
+      }
+
+      clearTimeout(requestTimeoutId)
+
+      if (this.current.response.status >= 500 && maybePerformRetry()) {
```

Comment:
This changes our retry behavior. Instead of retrying only on timeouts, we're also retrying on server errors – i.e. errors which are not related ot the request. Does that make sense?